File: libraries/dotnet/UrcConverter/Sources/Sm/Preprocess.fs
namespace UrcConverter.Sources.Sm

module internal Preprocess =

    open UrcConverter
    open UrcConverter.Sources.Sm.Model

    [<Literal>]
    let FastBpmWarp = 9999999.0

    let rowsOf (beats: float) : int =
        let value = beats * 48.0

        if value >= 0.0 then
            int (value + 0.5)
        else
            int (value - 0.5)

    /// Merges warp segments into [start, dest) row intervals; overlapping
    /// warps adopt the greater destination (StepMania semantics).
    let warpIntervals (warpSegs: (float * float) list) : (int * int) list =
        warpSegs
        |> List.map (fun (beat, length) -> let start = rowsOf beat in (start, start + rowsOf length))
        |> List.sortBy fst
        |> List.fold
            (fun (merged: (int * int) list) (start, dest) ->
                match merged with
                | (lastStart, lastDest) :: tail when start < lastDest ->
                    (lastStart, max lastDest dest) :: tail
                | _ -> (start, dest) :: merged)
            []
        |> List.rev

    /// Drops notes strictly inside a warp and truncates tails that end inside
    /// one to the warp start row.
    let filterNotes (notes: SmNote list) (intervals: (int * int) list) : SmNote list =
        notes
        |> List.filter (fun note ->
            intervals |> List.exists (fun (start, dest) -> start < note.Row && note.Row < dest)
            |> not)
        |> List.map (fun note ->
            match note.TailRow with
            | Some tailRow ->
                let truncated =
                    intervals
                    |> List.tryFind (fun (start, dest) -> start < tailRow && tailRow < dest)

                match truncated with
                | Some(start, _) -> { note with TailRow = Some start }
                | None -> note
            | None -> note)

    type private TimingEvent =
        | BpmEvt of beat: float * value: float
        | StopEvt of beat: float * value: float

    type private PreprocessState =
        {
            PrevBeat: float
            Bpm: float
            TimeOfs: float
            WarpStart: float
            PreWarpBpm: float
            OutBpm: (float * float) list
            OutStop: (float * float) list
            OutWarp: (float * float) list
        }

    /// Normalizes negative BPMs and negative stops into warps, matching
    /// StepMania's SMLoader::ProcessBPMsAndStops.
    let preprocess
        (timing: Timing)
        : Result<float * (float * float) list * (float * float) list * (float * float) list, UrcError> =
        let bpms = timing.Bpms |> List.sortBy fst
        let stops = timing.Stops |> List.sortBy fst

        let nonNegativeStops, initialOffset =
            stops
            |> List.fold
                (fun (accStops, accOffset) (beat, pause) ->
                    if beat < 0.0 then
                        (accStops, accOffset - pause)
                    else
                        ((beat, pause) :: accStops, accOffset))
                ([], timing.Offset)

        let sortedStops = nonNegativeStops |> List.rev

        let initialBpm, bpmRest =
            let beforeZero, afterZero = bpms |> List.partition (fun (beat, _) -> beat <= 0.0)

            let bpm =
                match List.tryLast beforeZero with
                | Some(_, value) -> value
                | None ->
                    match afterZero with
                    | (_, value) :: _ -> value
                    | [] -> 0.0

            bpm, afterZero

        if initialBpm = 0.0 then
            Error(UrcError.syntax 1 "no BPM in simfile")
        else
            let events =
                let bpmEvents = bpmRest |> List.map BpmEvt
                let stopEvents = sortedStops |> List.map StopEvt

                (bpmEvents @ stopEvents)
                |> List.sortWith (fun a b ->
                    let beatA, isBpmA =
                        match a with
                        | BpmEvt(b, _) -> b, true
                        | StopEvt(b, _) -> b, false

                    let beatB, isBpmB =
                        match b with
                        | BpmEvt(b, _) -> b, true
                        | StopEvt(b, _) -> b, false

                    if beatA <> beatB then
                        compare beatA beatB
                    elif isBpmA = isBpmB then
                        0
                    elif isBpmA then
                        -1
                    else
                        1)

            let initialState =
                {
                    PrevBeat = 0.0
                    Bpm = initialBpm
                    TimeOfs = 0.0
                    WarpStart = -1.0
                    PreWarpBpm = 0.0
                    OutBpm = if initialBpm > 0.0 && initialBpm <= FastBpmWarp then [ (0.0, initialBpm) ] else []
                    OutStop = []
                    OutWarp = []
                }

            let folder (state: PreprocessState) (event: TimingEvent) =
                let beat =
                    match event with
                    | BpmEvt(b, _)
                    | StopEvt(b, _) -> b

                let stateAfterTime =
                    if state.Bpm <= FastBpmWarp then
                        let nextTimeOfs = state.TimeOfs + ((beat - state.PrevBeat) * 60.0 / state.Bpm)

                        if state.WarpStart >= 0.0 && state.Bpm > 0.0 && nextTimeOfs > 0.0 then
                            let warpEnd = beat - (nextTimeOfs * state.Bpm / 60.0)
                            let outWarp = (state.WarpStart, warpEnd - state.WarpStart) :: state.OutWarp

                            let outBpm =
                                if state.Bpm <> state.PreWarpBpm then
                                    (state.WarpStart, state.Bpm) :: state.OutBpm
                                else
                                    state.OutBpm

                            { state with
                                PrevBeat = beat
                                TimeOfs = nextTimeOfs
                                WarpStart = -1.0
                                OutWarp = outWarp
                                OutBpm = outBpm
                            }
                        else
                            { state with
                                PrevBeat = beat
                                TimeOfs = nextTimeOfs
                            }
                    else
                        { state with PrevBeat = beat }

                match event with
                | BpmEvt(_, value) ->
                    if stateAfterTime.WarpStart < 0.0 && (value < 0.0 || value > FastBpmWarp) then
                        { stateAfterTime with
                            WarpStart = beat
                            PreWarpBpm = stateAfterTime.Bpm
                            TimeOfs = 0.0
                            Bpm = value
                        }
                    elif stateAfterTime.WarpStart < 0.0 then
                        { stateAfterTime with
                            Bpm = value
                            OutBpm = (beat, value) :: stateAfterTime.OutBpm
                        }
                    else
                        { stateAfterTime with Bpm = value }
                | StopEvt(_, value) ->
                    if stateAfterTime.WarpStart < 0.0 && value < 0.0 then
                        { stateAfterTime with
                            WarpStart = beat
                            PreWarpBpm = stateAfterTime.Bpm
                            TimeOfs = value
                        }
                    elif stateAfterTime.WarpStart < 0.0 then
                        { stateAfterTime with OutStop = (beat, value) :: stateAfterTime.OutStop }
                    else
                        let nextTimeOfs = stateAfterTime.TimeOfs + value

                        if value > 0.0 && nextTimeOfs > 0.0 then
                            let outWarp = (stateAfterTime.WarpStart, beat - stateAfterTime.WarpStart) :: stateAfterTime.OutWarp
                            let outStop = (beat, nextTimeOfs) :: stateAfterTime.OutStop

                            if stateAfterTime.Bpm < 0.0 || stateAfterTime.Bpm > FastBpmWarp then
                                { stateAfterTime with
                                    TimeOfs = 0.0
                                    WarpStart = beat
                                    OutWarp = outWarp
                                    OutStop = outStop
                                }
                            else
                                let outBpm =
                                    if stateAfterTime.Bpm <> stateAfterTime.PreWarpBpm then
                                        (stateAfterTime.WarpStart, stateAfterTime.Bpm) :: stateAfterTime.OutBpm
                                    else
                                        stateAfterTime.OutBpm

                                { stateAfterTime with
                                    TimeOfs = nextTimeOfs
                                    WarpStart = -1.0
                                    OutBpm = outBpm
                                    OutWarp = outWarp
                                    OutStop = outStop
                                }
                        else
                            { stateAfterTime with TimeOfs = nextTimeOfs }

            let finalState = events |> List.fold folder initialState

            let outBpm, outWarp =
                if finalState.WarpStart >= 0.0 then
                    let neverEnds = finalState.Bpm < 0.0 || finalState.Bpm > FastBpmWarp
                    let warpEnd =
                        if neverEnds then
                            99999999.0
                        else
                            finalState.PrevBeat - (finalState.TimeOfs * finalState.Bpm / 60.0)

                    let nextOutWarp = (finalState.WarpStart, warpEnd - finalState.WarpStart) :: finalState.OutWarp

                    let nextOutBpm =
                        if finalState.Bpm <> finalState.PreWarpBpm then
                            (finalState.WarpStart, finalState.Bpm) :: finalState.OutBpm
                        else
                            finalState.OutBpm

                    nextOutBpm, nextOutWarp
                else
                    finalState.OutBpm, finalState.OutWarp

            Ok(initialOffset, List.rev outBpm, List.rev finalState.OutStop, List.rev outWarp)

File: libraries/python/src/urc_converter/sources/sm/preprocess.py
"""SMLoader timing preprocessing and note filtering for StepMania simfiles."""

from ...error import UrcError
from .model import SmNote, Timing

_ROWS_PER_BEAT = 48.0
_FAST_BPM_WARP = 9999999.0


def rows(beats: float) -> int:
	"""Convert beat position to 48th-row index."""
	value = beats * _ROWS_PER_BEAT
	return int(value + 0.5) if value >= 0 else int(value - 0.5)


def warp_intervals(warp_segs: list[tuple[float, float]]) -> list[tuple[int, int]]:
	"""Merge warp segments into [start, dest) row intervals; adopt greater dest."""
	spans = sorted((rows(beat), rows(beat) + rows(length)) for beat, length in warp_segs)
	merged: list[list[int]] = []
	for start, dest in spans:
		if merged and start < merged[-1][1]:
			if dest > merged[-1][1]:
				merged[-1][1] = dest
		else:
			merged.append([start, dest])
	return [(start, dest) for start, dest in merged]


def filter_notes(notes: list[SmNote], intervals: list[tuple[int, int]]) -> list[SmNote]:
	"""Drop notes inside a warp and truncate tails that end inside one to warp start."""
	kept: list[SmNote] = []
	for note in notes:
		if any(start < note.row < dest for start, dest in intervals):
			continue
		if note.tail_row is not None:
			for start, dest in intervals:
				if start < note.tail_row < dest:
					note.tail_row = start
					break
		kept.append(note)
	return kept


def preprocess(
	timing: Timing,
) -> tuple[float, list[tuple[float, float]], list[tuple[float, float]], list[tuple[float, float]]]:
	"""Port of SMLoader::ProcessBPMsAndStops: normalize negative BPMs/stops into warps."""
	bpms = sorted(timing.bpms, key=lambda entry: entry[0])
	sorted_stops = sorted(timing.stops, key=lambda entry: entry[0])

	offset = timing.offset
	stops: list[tuple[float, float]] = []
	for beat, pause in sorted_stops:
		if beat < 0:
			offset -= pause
		else:
			stops.append((beat, pause))

	bpm = 0.0
	index = 0
	while index < len(bpms) and bpms[index][0] <= 0:
		bpm = bpms[index][1]
		index += 1
	if bpm == 0:
		if index == len(bpms):
			raise UrcError("syntax", 1, "no BPM in simfile")
		bpm = bpms[index][1]
		index += 1

	out_bpm: list[tuple[float, float]] = []
	out_stop: list[tuple[float, float]] = []
	out_warp: list[tuple[float, float]] = []
	if 0 < bpm <= _FAST_BPM_WARP:
		out_bpm.append((0.0, bpm))

	prevbeat = 0.0
	timeofs = 0.0
	warpstart = -1.0
	prewarpbpm = 0.0
	ibpm, istop = index, 0
	while ibpm < len(bpms) or istop < len(stops):
		change_is_bpm = istop >= len(stops) or (
			ibpm < len(bpms) and bpms[ibpm][0] <= stops[istop][0]
		)
		beat, value = bpms[ibpm] if change_is_bpm else stops[istop]

		if bpm <= _FAST_BPM_WARP:
			timeofs += (beat - prevbeat) * 60.0 / bpm
			if warpstart >= 0 and bpm > 0 and timeofs > 0:
				warpend = beat - (timeofs * bpm / 60.0)
				out_warp.append((warpstart, warpend - warpstart))
				if bpm != prewarpbpm:
					out_bpm.append((warpstart, bpm))
				warpstart = -1.0
		prevbeat = beat

		if change_is_bpm:
			if warpstart < 0 and (value < 0 or value > _FAST_BPM_WARP):
				warpstart = beat
				prewarpbpm = bpm
				timeofs = 0.0
			elif warpstart < 0:
				out_bpm.append((beat, value))
			bpm = value
			ibpm += 1
		else:
			if warpstart < 0 and value < 0:
				warpstart = beat
				prewarpbpm = bpm
				timeofs = value
			elif warpstart < 0:
				out_stop.append((beat, value))
			else:
				timeofs += value
				if value > 0 and timeofs > 0:
					out_warp.append((warpstart, beat - warpstart))
					out_stop.append((beat, timeofs))
					if bpm < 0 or bpm > _FAST_BPM_WARP:
						warpstart = beat
						timeofs = 0.0
					else:
						if bpm != prewarpbpm:
							out_bpm.append((warpstart, bpm))
						warpstart = -1.0
			istop += 1

	if warpstart >= 0:
		never_ends = bpm < 0 or bpm > _FAST_BPM_WARP
		warpend = 99999999.0 if never_ends else prevbeat - (timeofs * bpm / 60.0)
		out_warp.append((warpstart, warpend - warpstart))
		if bpm != prewarpbpm:
			out_bpm.append((warpstart, bpm))

	return offset, out_bpm, out_stop, out_warp
